logstash-core/lib/logstash/instrument/write_client.rb

```diff
@@ -0,0 +1,59 @@
+# encoding: utf-8
+module LogStash module Instrument
+  class WriteClient
+    def initialize(write_client, pipeline, metric, plugin)
+      @write_client = write_client
+
+      pipeline_id = pipeline.pipeline_id.to_s.to_sym
+      plugin_type = "#{plugin.class.plugin_type}s".to_sym
+
+      @events_metrics = metric.namespace([:stats, :events])
+      @pipeline_metrics = metric.namespace([:stats, :pipelines, pipeline_id, :events])
+      @plugin_metrics = metric.namespace([:stats, :pipelines, pipeline_id, :plugins, plugin_type, plugin.id.to_sym])
+
+      define_initial_metrics_values
+    end
+
+    def get_new_batch
+      @write_client.get_new_batch
+    end
+
+    def push(event)
+      record_metric(:push, event)
+    end
+    alias_method(:<<, :push)
+
+    def push_batch(batch)
+      record_metric(:push_batch, batch, batch.size)
+    end
+
+    private
+    def record_metric(method_name, arguments,  size = 1)
+      @events_metrics.increment(:in, size)
+      @pipeline_metrics.increment(:in, size)
+      @plugin_metrics.increment(:out, size)
+
+      clock = @events_metrics.time(:queue_push_duration_in_millis)
+
+      result = @write_client.send(method_name, arguments)
```

Comment:
I thought creating the proc would be slower than send since it have to keep local context? I need to benchmark that.